File: netCDF/hands_on/Solutions/F90/hands_on_par_read.f90
! This program reads data in parallel from a NetCDF file. 
! It is inspired by examples provided by NetCDF.
! Olga Abramkina (olga.abramkina@idris.fr), 2021

program hands_on_par_read 
  use netcdf
  implicit none
  include 'mpif.h'

  integer :: rank
  integer :: mpisize
  integer :: ierr
  
  character (len = *), parameter :: FILE_NAME = "par_write.nc"
  integer,parameter :: NDIMS=2
  integer,parameter :: NTIME=2

  integer :: ni_glo, nj_glo
  real, allocatable :: lon_glo(:),lat_glo(:)
  integer, allocatable :: field(:,:) 
  integer :: ni,ibegin,nj,jbegin
  integer :: i, j, t
  integer :: ncid, lat_dimid, lon_dimid, timedimid, varid, lonid, latid
  integer :: starts(NDIMS), counts(NDIMS)

  call MPI_Init(ierr)
  call MPI_Comm_rank(MPI_COMM_WORLD,rank,ierr)
  call MPI_Comm_size(MPI_COMM_WORLD,mpisize,ierr)

  ! Create the netcdf file.
  call check(nf90_open(FILE_NAME,NF90_NOWRITE,ncid,&
                         comm=MPI_COMM_WORLD,info=MPI_INFO_NULL))
 
  ! Get IDs of the two spacial dimensions given their names
  call check( nf90_inq_dimid(ncid, "lat", lat_dimid) )
  call check( nf90_inq_dimid(ncid, "lon", lon_dimid) )

  ! Get two spatial dimensions given their ids
  call check( nf90_inquire_dimension(ncid, lat_dimid, len = ni_glo) )
  call check( nf90_inquire_dimension(ncid, lon_dimid, len = nj_glo) ) 

  ! Assume that each MPI process reads the entirety of lat and lon
  allocate(lat_glo(ni_glo),lon_glo(nj_glo))

  ! Decompose global domain.
  ! Returned values: ni,nj,ibegin,jbegin.
  call decompose_domain(mpisize,rank,ni_glo,nj_glo,ni,ibegin,nj,jbegin)

  ! Contrary to lon, lat, assume that the varible is read in parallel
  ! Allocate local data array 
  allocate(field(ni,nj))

  ! Get IDs of two varialbes, latitude and longitude
  call check( nf90_inq_varid(ncid, "lat", latid) )
  call check( nf90_inq_varid(ncid, "lon", lonid) )

  ! Read latitude and longitude
  call check( nf90_get_var(ncid, latid, lat_glo) )
  call check( nf90_get_var(ncid, lonid, lon_glo) )

  ! Get ID of the temporal variable
  call check( nf90_inq_varid(ncid, "var_2d_temp", varid) )

  ! Set mode to collective for the temporal variable
  ! Not needed for reading
  !call check(nf90_var_par_access(ncid, varid, nf90_collective)) 

  ! Read the temporal variable
  do t=1,NTIME
    call check(nf90_get_var(ncid, varid, field, start=(/ibegin,jbegin,t/), count=(/ni,nj,1/)))

    ! Verify the read data
    do j = 1, nj
      do i = 1, ni
        if (field(i, j) /= (j-1+jbegin-1)*ni + (i-1+ibegin-1)) then
          print *, "Error: something went wrong; data read from the file do not correspond to the expected."
          stop 2
        endif
      end do
    end do

  enddo

  print *, "Parallel read is done successfully."

  ! Close file
  call check(nf90_close(ncid))

  deallocate(lat_glo, lon_glo, field)
  
  call MPI_FINALIZE(ierr)

  contains
  
  ! This subroutine part of the netCDF package.
  ! It handles errors by printing an error message
  ! and exiting with a non-zero status.
  subroutine check(status)
    integer,intent(in) :: status
    if(status /= nf90_noerr) then 
  !    print *, trim(nf90_strerror(status))
   print *, nf90_strerror(status)
      stop "Stopped"
    end if
  end subroutine check

  ! This subroutine decomposes global domain.
  subroutine decompose_domain(mpisize,rank,ni_glo,nj_glo,ni,ibegin,nj,jbegin)
    integer,intent(in) :: mpisize, rank, ni_glo, nj_glo
    integer,intent(out) :: ni, ibegin, nj, jbegin
    integer :: n
    ni=ni_glo ; ibegin=1
    jbegin=0
    do n=0,mpisize-1
      nj=nj_glo/mpisize
      if (n<MOD(nj_glo,mpisize)) nj=nj+1
      if (n==rank) exit
      jbegin=jbegin+nj
    enddo
    jbegin=jbegin+1
  end subroutine decompose_domain

END PROGRAM hands_on_par_read

File: netCDF/hands_on/C/hands_on_par_read.c
#include "mpi.h"
#include <netcdf.h>
#include <netcdf_par.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#define NDIMS     2
#define NTIME     2

#define ERR(e) {printf("Error: %s\n", nc_strerror(e)); return 2;}

// Function decomposes global domain of size ni_glo*nj_glo in horizontal direction into local domains
void decompose_domain(int mpisize, int mpirank, int ni_glo, int nj_glo, int* ni, int* ibegin, int* nj, int* jbegin)
{
    *nj=nj_glo;
    *ibegin=0;
    *jbegin=0;
    for (int n=0; n<mpisize; ++n)
    {	    
      *ni=ni_glo/mpisize;
      if (n<(ni_glo%mpisize)) ++(*ni);
      if (n==mpirank) break;	      
      (*ibegin)+=(*ni);
    }  
}

// Funciton checks the correctness of the read data for each time step
void check_read_data(int* data, int ni, int ibegin, int nj, int jbegin)
{
  int ok = 0;
  for(int i=0; i < ni; i++)
    for(int j=0; j < nj; j++)
      if (data[j+i*nj] != (j+jbegin) + (i+ibegin)*nj)
        ++ok;
  if (ok == 0)
    printf("Parallel read is done successfully.\n");
  else
    printf("Error: something went wrong; data read from the file do not correspond to the expected.\n");
}

int main (void)
{
  int mpirank, mpisize;
  /* Define variables holding netCDF ids of the file and three variables stored in it (longitude, latitude and a dummy 2D variable) */
  ...
  /* Define variables holding longitude and latitude */
  ...
  /* Define array holding three dimensions of the dummy 2D varialbe */
  ...
  /* Define start and count arrrays for parallel read of the variable */ 
  ...
  int err;
  size_t ni_glo, nj_glo;
  int ni, nj, ibegin, jbegin;
  char filename[128] = "par_write.nc";

  //--------------------------------------
  // Initialization MPI
  //--------------------------------------

  MPI_Init(NULL, NULL);
  MPI_Comm_rank(MPI_COMM_WORLD,&mpirank);
  MPI_Comm_size(MPI_COMM_WORLD,&mpisize);

  //--------------------------------------
  //  NetCDF metadata definitions
  //--------------------------------------

  /* Open a netcdf file */
  if (err=...) 
    ERR(err);

  /* Get IDs of two dimensions, latitude and longitude */
  if (err=...) 
    ERR(err);
  if (err=...) 
    ERR(err);

  /* Get sizes of latitude and longitude, store them into ni_glo, nj_glo */
  if (err=...) 
    ERR(err);
  if (err=...) 
    ERR(err);

  /* Assume that each MPI process reads the entirety of lat and lon */
  float *lat_glo = (float *)malloc(sizeof(float)*ni_glo);
  float *lon_glo = (float *)malloc(sizeof(float)*nj_glo);

  /* Get IDs of two varialbes, latitude and longitude */
  if (err=...) 
    ERR(err);
  if (err=...) 
    ERR(err);

  /* Read latitude and longitude from the file */
  if (err=...)
    ERR(err);
  if (err=...)
    ERR(err);

  /* Get ID of the temporal variable */
  if (err=...))
    ERR(...);

  /* Decompose global domain. Returned values: ni,nj,ibegin,jbegin */
  decompose_domain(mpisize,mpirank,ni_glo,nj_glo,&ni,&ibegin,&nj,&jbegin);
  int *data = (int *)malloc(sizeof(int)*ni*nj);
  
  //--------------------------------------
  // Read temporal data and close the file  
  //--------------------------------------

  // Define start and count arrays
  starts[0] = ...;  starts[1] = ...;  starts[2] = ...;
  counts[0] = ...;  counts[1] = ...;  counts[2] = ...;

  /* Read temporal data */
  for (int t=0; t<NTIME; ++t)
  { 
    starts[...] = ...;
    /* Read temporal variable */
    if (err=...) 
      ERR(err);
    /* Check the correctness of the read data */
    check_read_data(data, ni, ibegin, nj, jbegin);
  }

  /* Close the file */
  if (err=...) 
    ERR(err);

  free(data);
  free(lat_glo);
  free(lon_glo);

  MPI_Finalize();

  return 0;
}     

